File: generate_map.py
#!/usr/bin/env python3
"""
周年庆活动打卡地图生成器
将HTML地图转换为PNG图片
"""

import os
import time
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from PIL import Image
import io

def setup_chrome_driver():
    """设置Chrome浏览器驱动"""
    chrome_options = Options()
    chrome_options.add_argument('--headless')
    chrome_options.add_argument('--no-sandbox')
    chrome_options.add_argument('--disable-dev-shm-usage')
    chrome_options.add_argument('--disable-gpu')
    chrome_options.add_argument('--window-size=1920,1080')
    chrome_options.add_argument('--force-device-scale-factor=2')  # 提高分辨率
    
    # 尝试使用系统的Chrome/Chromium
    try:
        driver = webdriver.Chrome(options=chrome_options)
    except:
        # 如果失败，尝试使用chromium-browser
        chrome_options.binary_location = '/usr/bin/chromium-browser'
        driver = webdriver.Chrome(options=chrome_options)
    
    return driver

def html_to_png(html_file, output_file):
    """将HTML文件转换为PNG图片"""
    driver = None
    try:
        print("正在初始化浏览器...")
        driver = setup_chrome_driver()
        
        # 加载HTML文件
        file_url = f"file://{os.path.abspath(html_file)}"
        print(f"正在加载HTML文件: {file_url}")
        driver.get(file_url)
        
        # 等待页面完全加载
        time.sleep(3)
        
        # 获取地图容器元素
        map_element = driver.find_element(By.ID, "mapContainer")
        
        # 获取元素的位置和大小
        location = map_element.location
        size = map_element.size
        
        # 截取整个页面
        print("正在截取地图...")
        png = driver.get_screenshot_as_png()
        
        # 使用PIL裁剪图片到地图区域
        im = Image.open(io.BytesIO(png))
        
        # 由于设置了2倍的缩放因子，需要相应调整坐标
        left = location['x'] * 2
        top = location['y'] * 2
        right = left + size['width'] * 2
        bottom = top + size['height'] * 2
        
        # 裁剪图片
        im = im.crop((left, top, right, bottom))
        
        # 保存为PNG
        im.save(output_file, 'PNG', quality=95, dpi=(300, 300))
        print(f"地图已成功保存为: {output_file}")
        
        # 获取文件大小
        file_size = os.path.getsize(output_file) / 1024 / 1024
        print(f"文件大小: {file_size:.2f} MB")
        
    except Exception as e:
        print(f"转换过程中出现错误: {e}")
        raise
    finally:
        if driver:
            driver.quit()

def main():
    """主函数"""
    html_file = "/workspace/stamp_map.html"
    output_file = "/workspace/anniversary_stamp_map.png"
    
    print("=" * 50)
    print("周年庆活动打卡地图生成器")
    print("=" * 50)
    
    # 检查HTML文件是否存在
    if not os.path.exists(html_file):
        print(f"错误: HTML文件 {html_file} 不存在!")
        return
    
    # 转换为PNG
    html_to_png(html_file, output_file)
    
    print("=" * 50)
    print("地图生成完成！")
    print(f"输出文件: {output_file}")
    print("您可以下载此PNG文件用于打印或电子版使用")
    print("=" * 50)

if __name__ == "__main__":
    main()

File: generate_map.js
#!/usr/bin/env node

/**
 * 周年庆活动打卡地图生成器
 * 使用Puppeteer将HTML地图转换为PNG图片
 */

const puppeteer = require('puppeteer');
const fs = require('fs');
const path = require('path');

async function generateMapPNG() {
    console.log('='.repeat(50));
    console.log('周年庆活动打卡地图生成器');
    console.log('='.repeat(50));
    
    const htmlFile = path.join(__dirname, 'stamp_map.html');
    const outputFile = path.join(__dirname, 'anniversary_stamp_map.png');
    
    // 检查HTML文件是否存在
    if (!fs.existsSync(htmlFile)) {
        console.error(`错误: HTML文件 ${htmlFile} 不存在!`);
        return;
    }
    
    console.log('正在启动浏览器...');
    
    // 启动浏览器
    const browser = await puppeteer.launch({
        headless: 'new',
        args: [
            '--no-sandbox',
            '--disable-setuid-sandbox',
            '--disable-dev-shm-usage',
            '--disable-gpu'
        ]
    });
    
    try {
        const page = await browser.newPage();
        
        // 设置视口大小
        await page.setViewport({
            width: 1920,
            height: 1080,
            deviceScaleFactor: 2 // 提高分辨率
        });
        
        // 加载HTML文件
        const fileUrl = `file://${htmlFile}`;
        console.log(`正在加载HTML文件: ${fileUrl}`);
        await page.goto(fileUrl, {
            waitUntil: 'networkidle0',
            timeout: 30000
        });
        
        // 等待动画加载完成
        await page.waitForTimeout(3000);
        
        // 获取地图容器元素
        const element = await page.$('#mapContainer');
        
        if (!element) {
            throw new Error('找不到地图容器元素');
        }
        
        console.log('正在截取地图...');
        
        // 截取元素的屏幕截图
        await element.screenshot({
            path: outputFile,
            type: 'png'
        });
        
        // 获取文件大小
        const stats = fs.statSync(outputFile);
        const fileSizeMB = stats.size / (1024 * 1024);
        
        console.log(`地图已成功保存为: ${outputFile}`);
        console.log(`文件大小: ${fileSizeMB.toFixed(2)} MB`);
        console.log('='.repeat(50));
        console.log('地图生成完成！');
        console.log(`输出文件: ${outputFile}`);
        console.log('您可以下载此PNG文件用于打印或电子版使用');
        console.log('='.repeat(50));
        
    } catch (error) {
        console.error('生成过程中出现错误:', error);
    } finally {
        await browser.close();
    }
}

// 运行生成器
generateMapPNG().catch(console.error);